File: General-DSA-Questions/Queue/5_1stNonRepeatingCharInStream.cpp
// Write a function that takes a string s. While reading characters from the stream, you have to tell which character has appeared only once in the stream upto that point. If there are many characters that have appeared only once, you have to tell which one of them was the first one to appear. If there is no such character then append '#' to the answer.

// Appraoch :-
// 1. We will use a queue to store the characters that have appeared only once in the stream.
// 2. We will use a frequency array to store the frequency of each character in the stream.(Hashmap can also be used)
// 3. We will iterate over the string and do the following:-
//    a. Increment the frequency of the current character.
//    b. If the frequency of the current character is 1, we will push it into the queue.
//    c. If the frequency of the current character is greater than 1, we will pop the front element of the queue until we get a character whose frequency is 1.
//    d. If the queue is empty, we will append '#' to the answer.
//    e. Otherwise, we will append the front element of the queue to the answer.
// 4. Return the answer.
// Time Complexity :- O(n), because we are iterating over the string only once. Outer loop will run n times and inner loop will run at max n times. Since one character can be pushed and popped at most once, the inner loop will run at max n times.
// Total time complexity = O(n) + O(n) = O(n)
// Space Complexity :- O(1), because the frequency array will have at max 26 characters.

#include <iostream>
#include <queue>
#include <unordered_map>

using namespace std;

string firstNonRepeatingCharacter(string stream) {
    unordered_map<char, int> freq; // To store the frequency of characters
    queue<char> q;                // To maintain the order of characters
    string result = "";

    for (char ch : stream) {
        // Increment the frequency of the character
        freq[ch]++;

        // Add the character to the queue
        q.push(ch);

        // Remove characters from the front of the queue if they are repeating. i.e. freq > 1 or not present in the stream
        while (!q.empty() && freq[q.front()] > 1) {
            q.pop();
        }

        // Print the first non-repeating character or '#' if none exists
        if (!q.empty()) {
            result += q.front();
        } else {
            result += '#';
        }
    }
    return result;
}

int main() {
    string stream = "aabcdbcad";
    cout << "Stream: " << stream << endl;
    cout << "First non-repeating characters: " << firstNonRepeatingCharacter(stream) << endl;

    return 0;
}

File: JS-DSA-Questions/Queue/5_1stNonRepeatingCharInStream.js
// Given a stream of characters, for each character read tell which character has appeared only
// once in the stream up to that point. If several, report the first to appear. If none, append '#'.

// Example:
// Input: s = "aabcdbcad" → Output: "a#bbbcdd#"

// Approach:
// 1. Use a queue to store characters that appeared exactly once (in arrival order).
// 2. Use a frequency map for character counts.
// 3. For each character: increment its count, push it to the queue, then pop from the front while
//    the front character's count > 1. The front of the queue is the first non-repeating char.
// Time Complexity: O(n). Space Complexity: O(1) (at most 26 distinct chars).

// JavaScript note: C++ used std::queue<char>. Here a plain array with push()/shift() acts as the
// queue, and a Map (or object) holds the frequencies.
function firstNonRepeatingCharacter(stream) {
  const freq = new Map();
  const q = []; // queue of characters
  let result = "";

  for (const ch of stream) {
    freq.set(ch, (freq.get(ch) || 0) + 1);
    q.push(ch);

    // Remove repeating characters from the front
    while (q.length > 0 && freq.get(q[0]) > 1) {
      q.shift();
    }

    result += q.length > 0 ? q[0] : "#";
  }
  return result;
}

const stream = "aabcdbcad";
console.log("Stream:", stream);
console.log("First non-repeating characters:", firstNonRepeatingCharacter(stream)); // a#bbbcdd#
